tree/dataframe/inc/ROOT/RDFHelpers.hxx

```diff
@@ -181,6 +184,110 @@ RNode AsRNode(NodeType node)
 // clang-format on
 void RunGraphs(std::vector<RResultHandle> handles);
 
+
+// clang-format off
+/// RDF progress helper.
+/// This class provides callback functions to RDataFrame, which record event statistics
+/// and print them to the terminal every second.
+/// ProgressHelper::operator()(unsigned int, T&) is thread safe, and can be used as callback in MT mode.
+// clang-format on
+class ProgressHelper {
+public:
+
+  /// Create a progress helper.
+  /// \param increment RDF callbacks are called every `n` events. Pass this `n` here.
+  /// \param maxEvent If known, pass the number of events in the dataset. This will allow for computing completion
+  /// statistics.
+  /// \param progressBarWidth Number of characters the progress bar will occupy.
+  /// \param useShellColours Use shell colour codes to colour the output.
+  ProgressHelper(std::size_t increment,
+      std::size_t maxEvent,
+      unsigned int progressBarWidth = 40,
+      bool useShellColours = true) :
+  fMaxEvents{maxEvent},
+  fIncrement{increment},
+  fBarWidth{progressBarWidth},
+  fUseShellColours{useShellColours}
+  { }
+
+  ~ProgressHelper() = default;
+
+  // clang-format off
+  /// Thread-safe callback for RDataFrame.
+  /// It will record elapsed times and event statistics, and print a progress bar every second.
+  /// \param slot Ignored.
+  /// \param value Ignored.
+  // clang-format on
+  template<typename T>
+  void operator()(unsigned int /*slot*/, T& value) { operator()(value); }
+
+  // clang-format off
+  /// Thread-safe callback for RDataFrame.
+  /// It will record elapsed times and event statistics, and print a progress bar every second.
+  /// \param value Ignored.
+  // clang-format on
+  template<typename T>
+  void operator()(T& /*value*/) {
+    using namespace std::chrono;
+    // ***************************************************
+    // Warning: Here, everything needs to be thread safe:
+    // ***************************************************
+    fProcessedEvents += fIncrement;
+
+    // We only print every 1s, and only one thread does it. Try to take the lock:
+    if (duration_cast<seconds>(system_clock::now() - fLastPrintTime).count() == 0 || !fPrintMutex.try_lock())
```

Comment:
The access to `fLastPrintTime` and `fIncrement` seems thread unsafe?

Reply:
`const std::size_t fIncrement;`
I will say no more.

`fLastPrintTime` might indeed change, but the logic is this:
- If the time is right, try take a lock.
- If it's not right: exit.
- If it's right, but you cannot take a lock: exit.

Logic flawed? The progress bar behaved very nicely in my tests. Thread sanitizer might complain, though.

Reply:
> `const std::size_t fIncrement;` I will say no more.

:+1:

> Logic flawed?

LGTM, `RecordEvtCountAndTime` is only ever called from the locked region.

Reply:
Still since `fLastPrintTime` can be read unlocked (and thus written to at the same time), it probably should be made an `std::atomic`  (otherwise it can be spuriously be "right".